third/jwgc/clients/jlocate/jlocate.1
.\" Automatically generated by Pod::Man v1.34, Pod::Parser v1.13
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "JLOCATE 1"
.TH JLOCATE 1 "2003-10-06" "beta4" "Jabber Windowgram Client"
.SH "NAME"
jlocate \- display contacts and their current status
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
jlocate [ \-h ] [ \-d <flags> ] [ \-e ] [ \-a ] [ \-g ] [ \-n ] [ \-o ] [ accountname ]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fIJlocate\fR is used to display a list of contacts and their availability.
If an \fIaccountname\fR is specified, \fIjlocate\fR will check only accounts
that match the name you specify.  It tries to match as many accounts as
possible unless you specify the \fB\-e\fR option.  Otherwise, it performs it's
default behavior of retrieving the full contact list.  By default, this is
limited to only those that are online.  However, the various options
described below can alter the output style.
.PP
\&\fIJlocate\fR understands the following options:
.PP
\&\-h             Displays usage information.
.PP
\&\-d <flags>     Debugging output.  Run without <flags> for usage.
.PP
\&\-e             Exact match on account name (no effect if none specified).
.PP
\&\-a             Show full contact list (including unavailable and unknown).
.PP
\&\-g             Show contacts organized by group.
.PP
\&\-n             Swap nickname and jid positions in the listings.
.PP
\&\-o             Omit extra information (nickname or jid, dependent on \-n)
.SH "AUTHOR"
.IX Header "AUTHOR"
Daniel Henninger <jadestorm@users.sourceforge.net>
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright (c) 2002 Daniel Henninger. All right reserved.
This program is free software; you can redistribute it
and/or modify it under the terms explained in the \fB\s-1LICENSE\s0\fR
file that comes with the \fIjwgc\fR distribution.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIjwgc\fR\|(1), \fIjwrite\fR\|(1), \fIjctl\fR\|(1), \fIjstat\fR\|(1)


third/jwgc/clients/jlocate/jlocate.c
#include <netdb.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/socket.h>
#include <libjwgc.h>

int showall, groupmode, hideextra, swapjid, exactmatch;
char *target, *whoami;

void usage() {
	fprintf(stderr, "Usage: %s %s[-h] [-e] [-a] [-g] [-n] [-o] [-v] <account name>\n%s\
  -a   Show full contact list (normally unavailable is filtered)\n\
  -g   Show contacts by group\n\
  -e   Exact match on account name (no effect if none specified)\n\
  -h   Display help\n\
  -n   Show nickname on status line instead of jid\n\
  -o   Omit extra information (nickname or jid, depending on if -n is set)\n\
  -v   Verbose output\n\
\n\
  If no <account name> is provided, it returns your full contact list\n\
",
	whoami,
#ifdef NODEBUG
	"",
	""
#else
	"[-d <flags>] ",
	"  -d   Enable/Disable debugging (leave <flags> blank for usage)\n"
#endif /* NODEBUG */
	);
	exit(1);
}

void jlocate_standard_on_event_handler(jwgconn conn, jwgpacket packet) {
	xode x, r, rr;
	char *jid, *status, *nick, *resid;
	int cnt = 0;

	x = packet->x;
	x = xode_get_tag(x, "contact");
	for (;;) {
		if (!x) break;
		jid = xode_get_attrib(x, "jid");
		nick = xode_get_attrib(x, "nick");

		r = xode_get_tag(x, "resources");
		rr = xode_get_firstchild(r);
		if (!rr) {
			printf("%-49.49s %-15.15s %.16s\n",
				swapjid && nick ? nick : jid,
				"",
				"unavailable");
		}
		while (rr) {
			resid = xode_get_attrib(rr, "name");
			status = xode_get_attrib(rr, "status");
			printf("%-49.49s %-15.15s %.16s\n",
				swapjid && nick ? nick : jid,
				resid && strcmp(resid, "NULL") ? resid : "",
				status ? status : "unavailable");
			rr = xode_get_nextsibling(rr);
		}

		if (!hideextra) {
			if (swapjid && nick) {
				printf("    JID: %.69s\n", jid);
			}
			else if (!swapjid && nick) {
				printf("    Nickname: %.64s\n", nick);
			}
		}
		cnt++;
		x = xode_get_nextsibling(x);
	}

	printf("%d contacts found.\n", cnt);
}

void jlocate_grouped_on_event_handler(jwgconn conn, jwgpacket packet) {
	xode x, y, r, rr;
	char *jid, *status, *group, *nick, *resid;
	int cnt = 0;
	char *spacer;

	x = packet->x;
	x = xode_get_tag(x, "group");
	while (x) {
		group = xode_get_attrib(x, "name");
		if (group) {
			printf("%s:\n", strcmp(group,"") ? group : "Unknown");
			spacer = "    ";
		}
		else {
			spacer = "";
		}
		y = x;
		y = xode_get_tag(y, "contact");
		while (y) {
			jid = xode_get_attrib(y, "jid");
			nick = xode_get_attrib(y, "nick");

			r = xode_get_tag(y, "resources");
			rr = xode_get_firstchild(r);
			if (!rr) {
				printf("%s%-45.45s %-15.15s %.16s\n",
					spacer,
					swapjid && nick ? nick : jid,
					"",
					"unavailable");
			}
			while (rr) {
				resid = xode_get_attrib(rr, "name");
				status = xode_get_attrib(rr, "status");
				printf("%s%-45.45s %-15.15s %.16s\n",
					spacer,
					swapjid && nick ? nick : jid,
					resid && strcmp(resid, "NULL") ? resid : "",
					status ? status : "unavailable");
				rr = xode_get_nextsibling(rr);
			}

			if (!hideextra) {
				if (swapjid && nick) {
					printf("%s    JID: %.65s\n",
						spacer,
						jid);
				}
				else if (!swapjid && nick) {
					printf("%s    Nickname: %.60s\n",
						spacer,
						nick);
				}
			}
			cnt++;
			y = xode_get_nextsibling(y);
		}
		x = xode_get_nextsibling(x);
	}

	printf("%d contacts found.\n", cnt);
}

int main(argc, argv)
	int argc;
	char *argv[];
{
	int arg;
	xode x;
	jwgconn jwg;

	showall = hideextra = swapjid = groupmode = exactmatch = 0;
	target = NULL;

	whoami = argv[0];
	dinit();

	arg = 1;

	for (; arg < argc; arg++) {
		if (*argv[arg] != '-') {
			if (target) {
				usage();
			}
			target = argv[arg];
			showall = 1;
			continue;
		} 

		switch (argv[arg][1]) {
#ifndef NODEBUG
			case 'd':
				arg++;
				if (arg >= argc) {
					dprinttypes();
				}
				dparseflags(argv[arg]);
				break;
#endif /* NODEBUG */
			case 'a':
				showall = 1;
				break;
			case 'g':
				groupmode = 1;
				break;
			case 'h':
				usage();
				break;
			case 'e':
				exactmatch = 1;
				break;
			case 'n':
				swapjid = 1;
				break;
			case 'o':
				hideextra = 1;
				break;
			default:
				fprintf(stderr, "Unknown option: %s\n",
						argv[arg]);
				usage();
		}
	}

	jwg = jwg_new();
	if (!jwg) {
		fprintf(stderr, "jlocate: failed to initialize jwgc connection\n");
		exit(1);
	}
	if (groupmode) {
		jwg_event_handler(jwg, jlocate_grouped_on_event_handler);
	}
	else {
		jwg_event_handler(jwg, jlocate_standard_on_event_handler);
	}
	jwg_start(jwg);
	if (jwg_getfd(jwg) < 0) {
		fprintf(stderr, "jlocate: failed to create jwgc connection\n");
		exit(1);
	}
	x = xode_new("locate");
	if (target) {
		xode_put_attrib(x, "target", target);
		if (exactmatch) {
			xode_put_attrib(x, "match", "strict");
		}
		else {
			xode_put_attrib(x, "match", "loose");
		}
	}
	if (groupmode) {
		xode_put_attrib(x, "organization", "group");
	}
	if (showall) {
		xode_put_attrib(x, "showall", "yes");
	}
	jwg_send(jwg, x);
	xode_free(x);

	jwg_poll(jwg, -1);
	jwg_stop(jwg);

	exit(0);
}
